### ISSUE TITLE
stop publishing .asc.${checksums} to Central repository

### ISSUE THREAD

**Issue by hboutemy** [gradle/gradle#20232]
### Expected Behavior
`.asc` files are published to Maven Central without `.asc.${checksum}` 

### Current Behavior
for example: https://repo1.maven.org/maven2/org/junit/jupiter/junit-jupiter/5.8.2/
junit-jupiter-5.8.2-javadoc.jar.asc.md5
junit-jupiter-5.8.2-javadoc.jar.asc.sha1
junit-jupiter-5.8.2-javadoc.jar.asc.sha256
junit-jupiter-5.8.2-javadoc.jar.asc.sha512

and same for each `.asc`file

### Context
currently, for each artifact published to Central repository, there are many additional files published: checksums and signature
see for example: https://repo1.maven.org/maven2/org/junit/jupiter/junit-jupiter/5.8.2/

most of the files that I see follow expected Maven2 repository layout: https://maven.apache.org/repository/layout.html perfect

there is only one series of checksum files that I'm surprised of: each .asc file is currently complemented by checksum files .asc.md5+.asc.sha1+.asc.sha256+.asc.sha512

These files are not a Central requirement: see for example https://repo1.maven.org/maven2/org/codehaus/plexus/plexus-utils/3.4.1/

Is there a deliberate choice to add checksums for .asc files? Are these checksums found useful for something that I did not imagine?

I'd personally would think that these files should not be published.

### Steps to Reproduce 
publish a release to Maven Central

**Comment by vlsi** [comment]
I agree it makes no sense to have checksums for PGP signatures.

Do you have an official reference that mentions which files need checksums and which do not?

https://central.sonatype.org/publish/requirements/ does not seem to clarify the requirements regarding the checksums.

**Comment by hboutemy** [comment]
in fact, there is no reference for "what not to do" :)
but the reference is https://maven.apache.org/repository/layout.html for the layout

**Comment by vlsi** [comment]
@hboutemy , I am afraid you are slightly off here.
Maven is the tool, and Sonatype OSSRH (aka Central) is the repository that hosts artifacts. It is up to the repository to define rules and policies.
The tool should just obey the rules.

For instance, Maven as a tool can't really declare policies "who can publish artifacts under org.postgresql:postgresql coordinates". On the other hand, Central can declare rules like "every artifact must come with PGP signature in a form or `.asc` file, and the public key must exist on one of the ... keyservers".

That is why the Maven repository layout has absolutely no power here.

-----

Apparently, Central requires PGP signatures: https://central.sonatype.org/publish/requirements/#sign-files-with-gpgpgp

> All files deployed need to be signed with GPG/PGP and a .asc file containing the signature must be included for each file.

It has nothing to do with Maven, and there are repositories that do not really require PGP signatures when publishing artifacts.

That is why I believe, Gradle should not change the behavior at random, and Central repository should clarify exact requirements so everybody can follow.

---


Here's a real-life example where Cental artifacts miss PGP signatures:

https://repo.maven.apache.org/maven2/org/ow2/asm/asm/7.1/
The missing `.asc` files are caused by the fact that `asm` is published via synchronization from ow2 Nexus instance which does not really have `.asc` requirements.
https://gitlab.ow2.org/asm/asm/-/issues/317878

**Comment by hboutemy** [comment]
you're right, you're just ignoring the fact that the Maven tool and its `central` repository existed before the repository was externalized from Maven and its rules written with best efforts. I know that I point to Maven site, but the general repository layout has been defined by Maven (and works because it's quite extensible given classifiers and extensions are not predefined).

The way the rules are defined does not block additional unnecessary files that don't break the rules: for example 
`.asc.${checksum}`, or `.${checksum}.${checksum}` (= something that happens at Maven build tool level for reasons beyond this discussion).

it's only today that we're reading back the requirements more in-depth and finding some undocumented parts, when trying to fully abstract from its Maven origins: reading your feedback, I'm just discovering an obvious missing part in https://central.sonatype.org/publish/requirements/ = **there is nothing written about checksums** (lol!)

thinking at it, I know the reason: because `.md5` and `.sha1` were part of Maven2 repository layout requirement, then obvious in a Maven context

and it was "obvious" when writing

> All files deployed need to be signed with GPG/PGP and a .asc file containing the signature must be included for each file

that the `.asc` signature was not a "normal" file that was required to follow rules (= have checksums and a `.asc` signature file)

You're right that it should be documented: let's do it, open an OSSRH or MVNCENTRAL Jira issue and improve

Back to the Gradle publication plugin: just looking at an example https://repo.maven.apache.org/maven2/org/codehaus/plexus/plexus-utils/3.4.1/ shows that `.asc` files don't require checksums in real world

FYI, when looking for an example, I looked at Ant artifacts and found that the `.asc.${checksum}` files were also provided: https://repo.maven.apache.org/maven2/org/apache/ant/ant/1.10.9/ . Gradle is not the only tool suffering from the direct reading of the incomplete rules :)

**Comment by vlsi** [comment]
>https://central.sonatype.org/publish/requirements/ = there is nothing written about checksums (lol!)

Exactly. Central needs to publish the requirements in the first place.

---

https://issues.sonatype.org/browse/NEXUS-8213

It might be the case that Central behaviour regarding checksums for signatures has changed in 2013...

**Comment by donat** [comment]
Thanks for the report. It's not breaking anyone, so it's unlikely to be fixed in a timely fashion unless someone from the community provides a fix.

**Comment by vlsi** [comment]
For reference, here's the relevant code that publishes checksums: https://github.com/gradle/gradle/blob/8e0e8e442c411c212f305ac87a1df8c5b4bcc370/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publisher/AbstractMavenPublisher.java#L238-L244

I guess the hardest issue is to figure out how to make checksum publishing configurable.

I can easily imagine there might be maven repositories that require checksums for all the files, so issue is to design a proper configuration for the set of generated and published checksums.
For instance, someone might want to skip `md5`.

----

> It's not breaking anyone, so it's unlikely to be fixed

It does multiply the number of files in the repository which causes issue when adding sigstore signatures.

For instance, if you add sigstore signature for all the files, you add `.asc` and `.sha...` for all of them, so avoiding checksums for `.asc` is a clever way to keep the number of published files sane.

**Comment by marcphilipp** [comment]
> I agree it makes no sense to have checksums for PGP signatures.

I'm not convinced by that. Checksums exist to validate downloaded files have not been corrupted during transport which also applies to signature files. If a signature file got corrupted during transport, verifying the signature of the corresponding artifact would yield a misleading error message, would it not?

**Comment by hboutemy** [comment]
do you really have much transport corruption with https?

**Comment by aSemy** [comment]
I've made a project that demonstrates this.

https://github.com/aSemy/examples-gradle-signing-bugged

[signing-bug.zip](https://github.com/gradle/gradle/files/8839522/signing-bug.zip)


> > https://central.sonatype.org/publish/requirements/ = there is nothing written about checksums (lol!)
> 
> Exactly. Central needs to publish the requirements in the first place.
> 
> https://issues.sonatype.org/browse/NEXUS-8213
> 
> It might be the case that Central behaviour regarding checksums for signatures has changed in 2013...

The documentation says checksums aren't required for `*.asc` files, nor vice-versa.

> [Notice that .asc files don't need any additional checksum file, nor checksum files don't need
   .asc signature file.](https://central.sonatype.org/publish/requirements/#provide-files-checksums)

**Comment by vlsi** [comment]
I wonder what happends if the main artifact is named like `test.asc` where `asc` stands for `Android System Command`.

Would Central attempt to parse that `test.asc` as a PGP signature? Would it require the file to have `test.asc.asc` PGP signature?
Have anyone tried that?

**Comment by AbrahamArmasCordero** [comment]
any updates on this, kind of annoying to be honest

**Comment by hick209** [comment]
This is also annoying me. I'll push a PR to fix it.

**Comment by hick209** [comment]
There have been a couple of PRs on this, both were approved, but none ended up being merged.
Any particular reason for the hold up here?

**Comment by hick209** [comment]
@big-guy or @jvandort (from the PR), any idea when a fix for this (like #35247) would be merged in?

**Comment by illarionov** [comment]
Central is currently introducing monthly limits on the number of published files, making this issue even more important.

https://central.sonatype.org/publish/maven-central-publishing-limits/

**Comment by ljacomet** [comment]
We are aware of how this can impact the recently announced Maven Central publishing limits and are looking at options.

**Comment by ljacomet** [comment]
Fixed with:
* #38530
